lib/screens/config/cache.ex

```diff
@@ -0,0 +1,180 @@
+defmodule Screens.Config.Cache do
+  @moduledoc """
+  Functions to read data from a cached copy of the screens config.
+  """
+
+  alias Screens.Config
+  alias ScreensConfig.Devops
+
+  use Screens.Cache.Client, table: :screens_config
+
+  @type table_contents :: list(table_entry)
+
+  @type table_entry ::
+          {{:screen, screen_id :: String.t()}, ScreensConfig.Screen.t()}
+          | {:last_deploy_timestamp, DateTime.t()}
+          | {:devops, ScreensConfig.Devops.t()}
+
+  def ok?, do: table_exists?()
+
+  def refresh_if_loaded_before(screen_id) do
+    with_table default: nil do
+      [[last_deploy_timestamp]] = :ets.match(@table, {:last_deploy_timestamp, :"$1"})
+
+      refresh_if_loaded_before =
+        case :ets.match(@table, {screen_id, %{refresh_if_loaded_before: :"$1"}}) do
+          [[timestamp]] -> timestamp
+          [] -> nil
+        end
+
+      case {last_deploy_timestamp, refresh_if_loaded_before} do
+        {nil, nil} ->
+          nil
+
+        {nil, refresh_if_loaded_before} ->
+          refresh_if_loaded_before
+
+        {last_deploy_timestamp, nil} ->
+          last_deploy_timestamp
+
+        {last_deploy_timestamp, refresh_if_loaded_before} ->
+          Enum.max([last_deploy_timestamp, refresh_if_loaded_before], DateTime)
+      end
+    end
+  end
+
+  def last_deploy_timestamp do
+    with_table do
+      [[last_deploy_timestamp]] = :ets.match(@table, {:last_deploy_timestamp, :"$1"})
+      last_deploy_timestamp
+    end
+  end
+
+  def service_level(screen_id) do
+    with_table default: 1 do
```

Comment:
With this macro being called with a default of 1, is there a particular reason the `case` is needed? Is it possible for the `[]` case to match?

Reply:
Actually, I think I understand. Is the `[]` case there just in case the table does exist but the `screen_id` doesn't?

Reply:
Yup! There are two different failure modes for most of these functions that read from the cache.
1. Table doesn't exist (shouldn't happen, but...)
2. Table exists but no match is found for the given pattern (a bit more likely to happen)

Reply:
The `:default` value passed to `with_table`, or its fallback behavior of returning `{:error, "table #{x} doesn't exist"} handles (1), and the `[]` case handles (2).